File: Android/shiba/src/main/java/moe/tlaster/shiba/NativeRenderer.kt
package moe.tlaster.shiba

import android.content.Context
import moe.tlaster.shiba.dataBinding.ShibaBinding
import moe.tlaster.shiba.scripting.visitors.JSViewVisitor
import moe.tlaster.shiba.visitors.ValueVisitor
import org.liquidplayer.javascript.*

interface IShibaContext {
    fun getContext() : Context
    var dataContext: Any?
    val bindings: ArrayList<ShibaBinding>
    fun eventCallback(name: String)
}

internal object NativeRenderer {

    fun renderFromFunction(name: String, context: IShibaContext): NativeView {
        val creatorResult = Shiba.configuration.scriptRuntime.callFunction(name)
        if (creatorResult is JSValue) {
            val shibaView = JSViewVisitor.visit(creatorResult)
            if (shibaView is ShibaView) {
                return render(shibaView, context)
            }
        }
        return NativeView(context.getContext())
    }

    fun render(view: ShibaView?, context: IShibaContext): NativeView {
        return ValueVisitor.visit(view, context) as NativeView
    }
}

File: UWP/Shiba/NativeRenderer.cs
using ChakraHosting;
using Shiba.Controls;
using Shiba.Internal;
using Shiba.Scripting.Visitors;
using Shiba.Visitors;
using NativeView = Windows.UI.Xaml.UIElement;

namespace Shiba
{
    internal static class NativeRenderer
    {
        public static NativeView Render(View view, IShibaContext context)
        {
            return Singleton<ValueVisitor>.Instance.DynamicVisit(view, context) as NativeView;
        }

        public static NativeView RenderFromFunction(string name, IShibaContext context)
        {
            var result = ShibaApp.Instance.Configuration.ScriptRuntime.CallFunction(name);
            if (result is JavaScriptValue javaScriptValue)
                return Render(Singleton<JSViewVisitor>.Instance.Visit(javaScriptValue) as View, context);
            return null;
        }
    }
}